models/demos/wormhole/distilbert/tt/ttnn_optimized_distilbert.py

```diff
@@ -0,0 +1,419 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+import ttnn
+from typing import Optional
+import torch
+from ttnn.model_preprocessing import (
+    preprocess_linear_bias,
+    preprocess_linear_weight,
+)
+import ttnn.torch_tracer
+
+
+def get_head_mask(
+    head_mask: Optional[ttnn.Tensor],
+    num_hidden_layers: int,
+    is_attention_chunked: bool = False,
+):
+    head_mask = [
+        None,
+    ] * num_hidden_layers
+    return head_mask
+
+
+def attention(
+    config,
+    hidden_states,
+    mask,
+    head_mask=None,
+    output_attentions=None,
+    device=None,
+    base_address=None,
+    parameters=None,
+    num_cores_x=12,
+    min_val_tensor=None,
+    negative_val_tensor=None,
+    mesh_mapper=None,
+):
+    batch_size, q_length, dim = hidden_states.shape
+    k_length = hidden_states.shape[1]
+    dim_per_head = config.dim // config.n_heads
+
+    query_key_value_output = ttnn.linear(
+        hidden_states,
+        parameters.query_key_value.weight,
+        bias=parameters.query_key_value.bias,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat8_b,
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
+    )
+
+    (
+        query,
+        key,
+        value,
+    ) = ttnn.transformer.split_query_key_value_and_split_heads(
+        query_key_value_output,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        num_heads=config.n_heads,
+    )
+    ttnn.deallocate(query_key_value_output)
+
+    query = query * (1 / (dim_per_head) ** 0.5)
+
+    attention_scores = ttnn.matmul(
+        query,
+        key,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat16,
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
+    )
+    ttnn.deallocate(query)
+    ttnn.deallocate(key)
+    score_list = []
+
+    if batch_size <= 2:
+        inter_scores = attention_scores * negative_val_tensor
+        inter_scores = inter_scores + attention_scores
+        scores = inter_scores + min_val_tensor
+    else:
+        for i in range(2, batch_size + 1, 2):
```

Comment:
This looks very slow. For generative LMs we have added a scaled_dot_product_attention op that I think should be usable by distillbert too - see https://github.com/tenstorrent/tt-metal/blob/b52c50b24faac23195a9a603443dd9f5943902be/models/demos/wormhole/llama31_8b/tt/llama_attention.py#L444 for an example. If performance is a goal of this work then definitely look into this.

Reply:
This portion of the code is not related to the attention mechanism. It serves as a workaround for the masked_fill operation, as ttnn currently does not support masked_fill. To address this, a small workaround has been implemented. Since an Out Of Memory (OOM) error occurs for batch sizes greater than 2, the input tensor is sliced along axis 0, and the process is iterated within a loop to avoid memory issues.
Masked_fill issue: [#2795](https://github.com/tenstorrent/tt-metal/issues/2795)